### ISSUE TITLE
Setting up Azure AKS for CAP

### ISSUE THREAD

**Issue by svollath** [SUSE/doc-cap#68]
Just "saving" the commands here, so that they don't just live on some disks.

Below, the commands on which we should base our official documentation for Azure AKS.
Those are based on the blog from Thomas Conte (http://hypernephelist.com/2018/05/17/deploying-suse-cap-on-aks.html), but fine tuned and do work around https://github.com/Azure/AKS/issues/412.

```
# Prequisites
# kubectl, curl, sed, awk and jq commands


#
# Set parameters for your Azure AKS Cluster

# Set Resource Group name
export RGNAME="cap-aks"

# Set Azure Location
export REGION="westeurope"

# Set Kubernetes Node Count (min. 3 for CAP)
export NODECOUNT="3"

# Set Virtual Machine Size to use
export NODEVMSIZE="Standard_D2_v2"

# Set your public ssh key associated with your Azure account
export SSHKEYVALUE="~/.ssh/id_rsa.pub"


#
# Create the Resource Group

az group create --name $RGNAME --location $REGION


#
# Create the AKS Instance
# If you don't change the name, it will be the same as the "Resource Group"

export AKSNAME=$RGNAME

az aks create --resource-group $RGNAME --name $AKSNAME --node-count $NODECOUNT --admin-username "scf-admin" --ssh-key-value $SSHKEYVALUE --node-vm-size $NODEVMSIZE


#
# Retrieve the kubectl credentials

mv ~/.kube/config ~/.kube/config.old

az aks get-credentials --resource-group $RGNAME --name $AKSNAME


#
# Check kubernetes access

kubectl get nodes
kubectl get pods --all-namespaces


#
# Identify and set the Cluster Resource Group to work with

export MCRGNAME=$(az group list -o table | grep MC_"$RGNAME"_ | awk '{print$1}')


#
# Enable swap accounting

vmnodes=$(az vm list -g $MCRGNAME | jq -r '.[] | select (.tags.poolName | contains("node")) | .name')

for i in $vmnodes
do
   az vm run-command invoke -g $MCRGNAME -n $i --command-id RunShellScript --scripts "sudo sed -i 's|linux.*./boot/vmlinuz-.*|& swapaccount=1|' /boot/grub/grub.cfg"
done

for i in $vmnodes
do
   az vm restart -g $MCRGNAME -n $i
done


#
# Create and configure a Public IP and Load Balancer

az network public-ip create --resource-group $MCRGNAME --name $AKSNAME-public-ip --allocation-method Static

az network lb create --resource-group $MCRGNAME --name $AKSNAME-lb --public-ip-address $AKSNAME-public-ip --frontend-ip-name $AKSNAME-lb-front --backend-pool-name $AKSNAME-lb-back

nicnames=$(az network nic list --resource-group $MCRGNAME | jq -r '.[].name')

for i in $nicnames
do
    az network nic ip-config address-pool add \
    --resource-group $MCRGNAME \
    --nic-name $i \
    --ip-config-name ipconfig1 \
    --lb-name $AKSNAME-lb \
    --address-pool $AKSNAME-lb-back
done


#
# Configure load balancing and network security rules

# Relevant ports for CAP
export CAPPORTS="80 443 4443 2222 2793"

# Relevant ports for CAP incl. Stratos Web Console
export CAPPORTS="80 443 4443 2222 2793 8443"

for i in $CAPPORTS
do
    az network lb probe create --resource-group $MCRGNAME \
    --lb-name $AKSNAME-lb \
    --name probe-$i \
    --protocol tcp \
    --port $i

    az network lb rule create --resource-group $MCRGNAME \
    --lb-name $AKSNAME-lb \
    --name rule-$i \
    --protocol Tcp \
    --frontend-ip-name $AKSNAME-lb-front \
    --backend-pool-name $AKSNAME-lb-back \
    --frontend-port $i \
    --backend-port $i \
    --probe probe-$i
done

nsg=$(az network nsg list --resource-group=$MCRGNAME | jq -r '.[].name')
pri=200

for i in $CAPPORTS
do
    az network nsg rule create --resource-group $MCRGNAME \
    --priority $pri \
    --nsg-name $nsg \
    --name $AKSNAME-$i \
    --direction Inbound \
    --destination-port-ranges $i \
    --access Allow

    pri=$(expr $pri + 1)
done


#
# Print the public and private IPs for later use

echo -e "\n Resource Group:\t$RGNAME\n \
Public IP:\t\t$(az network public-ip show --resource-group $MCRGNAME --name $AKSNAME-public-ip --query ipAddress)\n \
Private IPs:\t\t\"$(az network nic list --resource-group $MCRGNAME | jq -r '.[].ipConfigurations[].privateIpAddress' | paste -s -d " " | sed -e 's/ /", "/g')\"\n"
```

Sample scf-config-values.yaml file for CAP on Azure AKS
```
secrets:
    # Password for user 'admin' in the cluster
    CLUSTER_ADMIN_PASSWORD: changeme

    # Password for SCF to authenticate with UAA
    UAA_ADMIN_CLIENT_SECRET: uaa-admin-client-secret

env:

    # Domain for SCF. DNS for *.DOMAIN must point to the a kube node's (not master)
    # external ip. This must match the value passed to the
    # cert-generator.sh script.
    DOMAIN: <Public IP>.xip.io
    # UAA host/port that SCF will talk to. If you have a custom UAA
    # provide its host and port here. If you are using the UAA that comes
    # with the SCF distribution, simply use the two values below and
    # substitute the cf-dev.io for your DOMAIN used above.
    UAA_HOST: uaa.<Public IP>.xip.io
    UAA_PORT: 2793
    #Azure deployment requires overlay
    GARDEN_ROOTFS_DRIVER: "overlay-xfs"
kube:
    # The IP address assigned to the kube node pointed to by the domain. 
    external_ips: [<Private IPs>]
    storage_class:
        # Make sure to change the value in here to whatever storage class you use
        persistent: "default"
        shared: "shared"
    # The registry the images will be fetched from. The values below should work for
    # a default installation from the suse registry.
    registry:
       hostname: "registry.suse.com"
       username: ""
       password: ""
    organization: "cap"

    auth: none
```

**Comment by mook-as** [comment]
With the changes in https://github.com/SUSE/fissile/pull/359 (i.e. estimated next release of SCF/CAP), we won't need to touch load balancers manually anymore. So anything with `$AKSNAME-lb` in it will be unncessary.  I do not know if we still need the manually-created public IPs and things like that; I followed the guide above, then had to delete the created load balancers before things worked.

I _did_ have to manually set up DNS entries for those automatically created load balancers, though :unamused: 

Additionally, `kube.external_ips` should not be set in that case.

**Comment by cjschroder** [comment]
Hi @mook-as, I haz questions. Because I've gone through the steps several times, with success, and this is my final version: http://docserv.nue.suse.com/documents/CAP_1/cap-deployment/single-html/#cha.cap.depl-azure

What SCF/UAA version are you deploying? I'm using the 2.8.0 Helm charts, and 1.1.0 for Stratos.

This load balancer business is nutty. Kubernetes has its own internal load balancer, and then external load balancers can serve a number of purposes. What will the new CAP load balancer do?

How did you manage your DNS, did you use Azure's DNS service?

**Comment by cjschroder** [comment]
additional resource: http://hypernephelist.com/2018/05/17/deploying-suse-cap-on-aks.html

**Comment by mook-as** [comment]
I'm on `develop`, which would be 1.2 at some point :D That comment was meant to be forward-looking, since we've historically been terrible at telling you things before they happen.  I guess I went too far the other way instead :grin: 

The new load balancers are the kubernetes ones; we're just going to be smarter and actually use those instead of having custom user-deployed ones that go through a single (kubernetes worker) node.

And yes, I did use Azure's DNS service; they aren't directly tied into the kubernetes load balancers, though, so any DNS should work fine.
